app/services/chat_completion_a_p_i.rb

```diff
@@ -0,0 +1,154 @@
+class ChatCompletionAPI
+
+  # This is a lightweight wrapper around the OpenAI::Client gem that hides away a bunch of the complexity.
+  #
+  # ChatCompletionAPI.get_next_response("You are a comedian", ["Tell me a joke"], model: "gpt-4")
+  #
+
+  def self.api_key
+    Rails.env.test? ? "" : Current.user.openai_key
+  end
+
+  def self.get_next_response(system_message, chat_messages, params = {})
+    message_payload = [{
+      role: 'system',
+      content: system_message
+    }]
+
+    chat_messages.each_with_index do |msg, i|
+      role = (i % 2 == 0) ? 'user' : 'assistant'
+
+      message_payload << {
+        role: role,
+        content: msg
+      }
+    end
+
+    response = call_api({messages: message_payload}.merge(params)) # format: https://platform.openai.com/docs/api-reference/chat
+  end
+
+
+  private
+
+  def self.model_or_engine
+    "gpt-3.5-turbo-1106"
+  end
+
+  def self.url
+    "https://api.openai.com/v1/chat/completions"
+  end
+
+  def self.default_params
+    {
+      model: model_or_engine,
+      max_tokens: 500,                      # a sensible default
+      n: 1,
+      response_format: { "type": "text" },  # or json_object
+    }
+  end
+
+  def self.call_api(params)
+    params = default_params.deep_symbolize_keys.merge(params.symbolize_keys)
+    verify_params!(params)
+
+    client = OpenAI::Client.new(
+      access_token: api_key,
+      request_timeout: 240,
+    )
+
+    response = ""
+    retries = 0
+    keep_retrying = true
+
+    begin
+      verify_token_count!(params)
+      response = ""
+      finished_reason = nil
+
+      if Rails.env.test?
+        raw_response = client.chat(parameters: params)
+        response = JSON.parse(raw_response)&.dig("choices", 0, "message", "content")
+      else
+        client.chat(parameters: params.merge(stream: proc { |chunk, _bytesize|
+          finished_reason = chunk&.dig("choices", 0, "finish_reason")
+
+          if !finished_reason
+            response += chunk&.dig("choices", 0, "delta", "content")&.to_s
+          else
+            raise finished_reason  unless finished_reason == "stop"
+          end
+        }))
+      end
+    rescue => e
+      if retries < 2 && e.class == RestClient::Exceptions::ReadTimeout
+        retries = retries + 1
+        sleep (1*retries)
+        retry
+      elsif retries < 2 && e.message == 'length' && (content = params[:messages].last[:content]).length > 100000
+        retries = retries + 1
+        sleep (1*retries)
+
+        params[:messages][-1][:content] = content[0...50000] + "\n...\n" + content[-50000...]
+        retry
+      else
+        puts "Error: retried #{retries} times. Error: #{e}"
+        #binding.pry  unless e.message == 'length'
+        sleep 2
+        retry  if keep_retrying
+      end
+    end
```

Comment:
This rescue block makes me uncomfortable, but I know some of it is forged with hard won lessons. Here are some things that make me skeptical:

- This rescues every exception, and will blindly retry even when exceptions occur that have nothing to do with the http api code above. If someone writes a bug in lines 72-79 it'll retry even a successful API request, burning API credits in the process. Having a dedicated `rescue RestClient::Exceptions::ReadTimeout =>` clause would draw the scope in a bit.
- The retry counting logic is muddled with stuff which checks the exception message for an arbitrary "length" string and checks the length of some parameter...and then changes the parameter? There's some kind of heuristic going on here which implicitly represented, but I think if it were explicit it would be easier to understand. Would it be possible to separate the "retry the request when it fails and needs to be restructured" logic out into a different place on the stack?
- The raise on line 78 is also opaque, and I think it's triggering some of this logic somehow. Is it possible to represent that logic without raising a string and catching it? It feels like throwing an exception here doesn't represent an exceptional circumstance, but an ordinary one.

Reply:
@robacarp yea, good call.I copied this library from another project and the rescue block was actually twice as complex. :) I stripped out a lot of project-specific stuff but you're right that even what I left is pretty heavy handed.

I'll just remove the rescue for now and we'll re-build up checks as we actually need them, for this project.